src/coreclr/jit/utils.h

```diff
@@ -773,4 +775,138 @@ int64_t GetSigned64Magic(int64_t d, int* shift /*out*/);
 
 double CachedCyclesPerSecond();
 
+namespace CheckedOps
+{
+const bool Unsigned = true;
+const bool Signed   = false;
+
+// Important note: templated functions below must use dynamic "assert"s instead of "static_assert"s
+// because they can be instantiated on code paths that are not reachable at runtime, but visible
+// to the compiler. One example is VN's EvalOp<T> function, which can be instantiated with "size_t"
+// for some operators, and that's legal, but its callee EvalOpSpecialized<T> uses "assert(!AddOverflows(v1, v2))"
+// for VNF_ADD_OVF/UN, and would like to continue doing so without casts.
+
+template <class T>
+bool AddOverflows(T firstAddend, T secondAddend, bool unsignedAdd)
+{
+    typedef typename std::make_unsigned<T>::type UT;
+    assert((std::is_same<T, int32_t>::value || std::is_same<T, int64_t>::value));
+
+    T result = firstAddend + secondAddend;
+
+    // For the SIGNED case - If there is one positive and one negative operand, there can be no overflow.
+    // If both are positive, the result has to be positive, and similary for negatives.
+    // For the UNSIGNED case - If an unsigned operand is bigger than the result then OVF.
+    if (unsignedAdd)
+    {
+        if ((static_cast<UT>(firstAddend) > static_cast<UT>(result)) ||
+            (static_cast<UT>(secondAddend) > static_cast<UT>(result)))
+        {
+            return true;
+        }
+    }
+    else if (((firstAddend < 0) == (secondAddend < 0)) && ((firstAddend < 0) != (result < 0)))
+    {
+        return true;
+    }
+
+    return false;
+}
+
+template <class T>
+bool SubOverflows(T minuend, T subtrahend, bool unsignedSub)
+{
+    typedef typename std::make_unsigned<T>::type UT;
+    assert((std::is_same<T, int32_t>::value || std::is_same<T, int64_t>::value));
+
+    T result = minuend - subtrahend;
+
+    if (unsignedSub)
+    {
+        if (static_cast<UT>(subtrahend) > static_cast<UT>(minuend))
+        {
+            return true;
+        }
+    }
+    else
+    {
+        // If both operands are positive or negative, there can be no overflow.
+        if ((minuend < 0) != (subtrahend < 0))
+        {
+            // If the signs of the operands differ and there was no overflow,
+            // the result will always match the sign of minuend.
+            // Consider A > 0, B > 0:
+            // sign((-A) - (+B)) == sign(-A - B) == -1 == sign(-A).
+            // sign((+A) - (-B)) == sign(A + B) == 1 == sign(A).
```

Comment:
These lines are a bit confusing as the `sign` function has three values `-1`, `0`, and `1`, while the code checks for `x  < 0` only without distinguishing zero and strictly positive values.  I think this part of the comment may be safely removed.

Reply:
Too addressed by using `ClrSafeInt<T>`.